system/Image/Image.php

```diff
@@ -0,0 +1,163 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @package	CodeIgniter
+ * @author	CodeIgniter Dev Team
+ * @copyright	Copyright (c) 2014 - 2016, British Columbia Institute of Technology (http://bcit.ca/)
+ * @license	http://opensource.org/licenses/MIT	MIT License
+ * @link	http://codeigniter.com
+ * @since	Version 3.0.0
+ * @filesource
+ */
+
+/**
+ * Image manipulation factory.
+ *
+ * Creates and returns an instance of the appropriate image manipulation processor.
+ *
+ * @package CodeIgniter\Image
+ */
+class Image
+{
+    protected $driver;
+    protected $resource;
+    protected $imageFullPath;
+    protected $imageFolder;
+    protected $imageFilename;
+    protected $imageExtension;  
+    protected $mime;
+    
+    protected $imageTypes = [
+            'jpg',
+            'jpeg',
+            'png',
+            'gif',
+            'tiff'
+    ];
+    
+    // ------------------------------------------------------------------
+    
+    /**
+     * @param string $driver
+     * @return resource
+     */
+    public function load(string $driver): resource
+    {
+        $classname = "Codeigniter\\Image\\$driver";
+        return new $classname;
+    }
+    
+    // ------------------------------------------------------------------
+    
+    /**
+     * @param string $folder
+     * @return resource
+     */
+    public function setFolder(string $folder): resource
+    {
+        $this->imageFolder = $folder;
+        return $this;
+    }
+    
+    // ------------------------------------------------------------------
+    
+    /**
+     * @param string $filename
+     * @return resource
+     */
+    public function setFilename(string $filename): resource
+    {
+        $this->imageFilename = $filename;
+        return $this;
+    }
+    
+    // ------------------------------------------------------------------
+    
+    /**
+     * @param string $extension
+     * @throws \InvalidArgumentException
+     * @return resource
+     */
+    public function setExtension(string $extension): resource
+    {
+        if (in_array($extension, $this->imageTypes))
+        {
+            $this->imageExtension = strtolower($extension);
+            return $this;
+        }
+        else
+        {
+            throw new \InvalidArgumentException('Image extension is not allowed');
+        }
+    }
+    
+    // ------------------------------------------------------------------
+    
+    /**
+     * @param string $full_image_path
+     * @throws \InvalidArgumentException
+     * @return resource
+     */
+    public function image(string $full_image_path = false, $driver): resource
```

Comment:
The naming of this method is a little awkward, `Image::image()`, and doesn't really specify whether we're creating a new empty image to work with, reading one from file, etc. And your docblock doesn't help out there either. :)

Perhaps it would be more appropriately named `fromFile()`?